test: Reduce build durations

pchaigno opened the pull request:
Summary of commits:
1. Adds some helper functions to help in subsequent commits.
3. Limits K8sHubble to run on GKE and 4.9 only.
2. Limits K8sIdentity and K8sHealth to run on GKE only.
4. and 5. Avoid installing Cilium for tests that are skipped or quarantined.

pchaigno committed:
test: Helpers to check if running on GKE

And make use of the new RunsOnGKE/DoesNotRunOnGKE helpers in existing
testing code.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
test: Avoid installing Cilium for K8sConformance if it is skipped

In K8sConformance, we skip the tests inside the It blocks instead of
using SkipIfIf. Because of that, ginkgo executes the
BeforeAll/BeforeEach blocks before figuring out the It blocks are
skipped anyway.

This commit fixes it by using SkipContextIf with all It blocks inside.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
test: Avoid installing Cilium for K8sCLI if quarantined

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
test: Run K8sIdentity and K8sHealth tests on GKE only

These tests should not be impacted by kernel versions and are therefore
safe to test only on GKE. The three test suites also needed to be
reorganized to avoid installing Cilium when the tests are skipped. To
that effect, all tests are moved into a new Context block.

Signed-off-by: Paul Chaignon <paul@cilium.io>

kkourt commented on test/k8sT/DatapathConfiguration.go:
```diff
@@ -288,15 +286,8 @@ var _ = Describe("K8sDatapathConfig", func() {
-		})
-
+	// DirectRouting without AutoDirectNodeRoutes not supported.
+	SkipContextIf(helpers.DoesNotRunOnGKE, "DirectRouting", func() {
```
nit: "not supported on GKE."

pchaigno replied:
Updated to `not supported outside of GKE.`

kkourt commented on test/k8sT/hubble.go:
```diff
@@ -32,149 +32,147 @@ import (
 var _ = Describe("K8sHubbleTest", func() {
+	// We want to run Hubble tests both with and without our kube-proxy
+	// replacement, as the trace events depend on it. We thus run the tests
+	// on GKE and our 4.9 pipeline.
```
Comment says 4.9 and GKE but we seem to check for not running on NetNext or or 4.19 (`RunsOnNetNextOr419Kernel`) which, I guess, is the complement set. Maybe that's how we do this thing, but it can be somewhat confusing and might have unexpected results if we add kernels.

pchaigno replied:
I was going to make that change, but after some thoughts, I'm not sure it makes sense:
We don't have a `RunsOn49Kernel()` helper because we don't pass a flag for 4.9. 4.9 in our CI is literally defined as the lack of either `KERNEL=419` or `NETNEXT=1`. So RunsOn49Kernel() would really be `DoesNotRunOnNetNextOr419Kernel()`. With the GKE special case, that would bring us to write:
```go
SkipContextIf(func() bool {
	return !helpers.RunsOnGKE() && helpers.RunsOnNetNextOr419Kernel()
}, ...
```
which is exactly equivalent to:
```go
SkipContextIf(helpers.RunsOnNetNextOr419Kernel, ...
```

When we add new kernels (e.g., `KERNEL=54`) I think we'll have to rework that set of helpers anyway.

kkourt replied:
I don't understand that infrastructure that well, and I think leaving it as it is is perfectly fine since the comment above explains it.

Tangent: as a user, I would love to avoid using env variables (especially, differently named ones) at the ginkgo level and just pass options. I haven't had a chance to look how easy the conversion would be, but if we ever decide to add another kernel, switching to options (e.g., cilium.kernel or cilium.environment) instead of env vars might be something to consider.

pchaigno replied:
> Tangent: as a user, I would love to avoid using env variables (especially, differently named ones) at the ginkgo level and just pass options. I haven't had a chance to look how easy the conversion would be

I agree. It should be fairly trivial to implement by following [one of our existing options](https://github.com/cilium/cilium/blob/e8605f88baabceebef52170b7337c793436492aa/test/config/config.go#L33). Just needs someone to invest cycles.

kkourt reviewed (approved):
Just two minor nits. I don't think is worth redoing the PR for them.

pchaigno committed:
test: Run K8sHubble on GKE and 4.9 pipelines

The Hubble tests check packet trace events that change depending on
whether our kube-proxy replacement is used. We thus run those tests both
on GKE (with our kube-proxy replacement) and on 4.9 (without). We
however don't need to run on all our kernel versions.

This commit only moves all the initialization code into the Hubble
Observe Context, merging BeforeAlls, AfterAll, and variables.

Signed-off-by: Paul Chaignon <paul@cilium.io>

gandro reviewed (approved):
Hubble related changes look good to me!
